### ISSUE TITLE
Converting an empty dense tensor to sparse_coo makes an uncoalesced tensor

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#73426]
### 🐛 Describe the bug

```shell
python -c 'import torch; torch.tensor([]).to_sparse().indices()'
# RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first
```

This error is confusing. It should either return an empty indices tensor (which makes sense) or return an error about an empty tensor not supported

not sure if this reproduces on newer releases

### Versions

1.9.1

cc @nikitaved @pearu @cpuhrsch

**Comment by cpuhrsch** [comment]
@vadimkantorov - Thank you for submitting the issue! Does this hold true on the latest nightly?

**Comment by vadimkantorov** [comment]
don't have time to check right now :( i can check in two weeks

**Comment by ezyang** [comment]
This is still broken

**Comment by vadimkantorov** [comment]
I wonder, maybe modifying is_coalesced() directly to return True for empty tensors (instead of actually caring to set the flag) is also a solution? not sure if it's helpful

**Comment by ezyang** [comment]
Yeah, I'm not sure what the best way to hit all the spots where we're doing the invariants wrong is. Another strategy is to get the SparseTensorImpl to default this field when nnz is zero. cc @bhosmer

**Comment by bhosmer** [comment]
> Yeah, I'm not sure what the best way to hit all the spots where we're doing the invariants wrong is. Another strategy is to get the SparseTensorImpl to default this field when nnz is zero. cc @bhosmer

This solution seems like the right one to me. (Not sure whether "all the spots where we're doing the invariants wrong" are, would they be solved by this?)

@amjames @nikitaved would one of you be the right person to ask to do this? also cc @cpuhrsch ... _for now_

**Comment by amjames** [comment]
This particular issue is due to the early exit for the nnz case in `dense_to_sparse` which does not set coalesced (see the early exit on line 527 vs the "real" one on 558). 

https://github.com/pytorch/pytorch/blob/8707aabe9af60645a1d0b0a85ba9a0167a28a1bd/aten/src/ATen/native/sparse/SparseTensor.cpp#L526-L558

Having SparseTensorImpl default the field when nnz is zero would help but we should probably be careful in review to consider the `coalesced` state of the return of an op. And ensure that any path along which it can be guaranteed it is set. However doing that retroactively for everything that is already implemented is not a small amount of work.

I can handle this issue, and the shortcut for `nnz == 0` ( should really be nnz < 2, as if you only have one nnz it must be coalesced also) in `is_coalesced()`

**Comment by pearu** [comment]
> `# RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first`
> This error is confusing. It should either return an empty indices tensor (which makes sense) or return an error about an empty tensor not supported

Formally, the error message is correct and it is not specific to `to_sparse()` operation. For example, `(a - a).indices()` would produce the same error when `a` is an uncoalesced tensor while one could expect that the result ought to be `a.indices()` (or even `tensor([], dtype=int64)`).

Recall that uncoalesced tensor is an optimization technique that avoids expensive coalesce operation for additive operations when performed subsequently multiple times. Uncoalesced and coalesced tensors are mathematically equivalent and equal after applying `.coalesce()` method.

What could be confusing here is the fact that it is not well-defined when an operation ought to return an uncoalesced or a coalesced tensor. Although [docs](https://pytorch.org/docs/stable/sparse.html#uncoalesced-sparse-coo-tensors) state that one "shouldn’t have to care whether or not a sparse tensor is coalesced or not", this issue illustrates that this statement could be over-optimistic: while most non-additive operations coalesce their inputs automatically then `indices()` and `values()` don't for certain reasons (and these methods represent data access operations rather than operations on tensors).

Let me propose the following criteria for what to expect when dealing with COO tensors as inputs and outputs for various PyTorch operations:
1. all non-additive operations always return a coalesced tensor, if supported, because these coalesce inputs automatically.
2. if input(s) to an additive operation is a coalesced tensor, the output is guaranteed to be a coalesced tensor as well. Here we define that strided as well as CSR/CSC/BSR/BSC tensors are coalesced tensors by nature.
3. if any input to an additive operation is an uncoalesced tensor, the output may or may not be an uncoalesced tensor. Implementations are encouraged to return a coalesced tensor when coalesced or uncoalesced tensor outputs would be equivalent performance-wise.

Then, according to criteria 2, this issue is valid: `to_sparse()` ought to return coalesced tensor for all strided inputs, not just for an empty dense tensor that has zero dimensions.

As observed in #82426 test failures, there exist operations that would fail on the criteria 2 for certain inputs if we would define all COO tensors with `nnz < 2` be as coalesced tensors (as the criteria 3 would recommend). So, this issue may have wider implications.

**Comment by vadimkantorov** [comment]
Also, in theory, `indices()`/`values()` could just return uncoalesced underlying tensors if the tensor isn't coalesced

**Comment by pearu** [comment]
The issue appears to be fixed via #82426
```python
>>> import torch; torch.tensor([]).to_sparse().indices()
tensor([], size=(1, 0), dtype=torch.int64)
```
So, closing.

**Comment by vadimkantorov** [comment]
> It can happen (e.g. in `a - a` case), that certain cells could turn into zeros (and then what is the semantics? they get removed from the sparse tensor? or they would still be in indices, just the value would be zero? are there any differences wrt gradient computation between a non-existent value and existing zero value)

curious about this
